src/qt/treectrl.cpp

```diff
@@ -56,275 +495,699 @@ void wxTreeCtrl::SetIndent(unsigned int indent)
+        : wxTreeItemId();
 }
 
 wxTreeItemId wxTreeCtrl::GetFirstVisibleItem() const
 {
+   wxTreeItemId itemid = GetRootItem();
+    if ( !itemid.IsOk() )
+        return itemid;
+
+    do
+    {
+        if (IsVisible(itemid))
+            return itemid;
+        itemid = GetNext(itemid);
+    } while ( itemid.IsOk() );
+
     return wxTreeItemId();
 }
 
 wxTreeItemId wxTreeCtrl::GetNextVisible(const wxTreeItemId& item) const
 {
+    wxCHECK_MSG(item.IsOk(), wxTreeItemId(), wxT("invalid tree item"));
+    wxASSERT_MSG(IsVisible(item), wxT("this item itself should be visible"));
+
+    wxTreeItemId id = item;
+    if ( id.IsOk() )
+    {
+        while ( id = GetNext(id), id.IsOk() )
+        {
+            if ( IsVisible(id) )
+                return id;
+        }
+    }
     return wxTreeItemId();
 }
 
 wxTreeItemId wxTreeCtrl::GetPrevVisible(const wxTreeItemId& item) const
 {
-    return wxTreeItemId();
+    wxCHECK_MSG(item.IsOk(), wxTreeItemId(), wxT("invalid tree item"));
+    wxASSERT_MSG(IsVisible(item), wxT("this item itself should be visible"));
+
+    // find out the starting point
+    wxTreeItemId prevItem = GetPrevSibling(item);
+    if ( !prevItem.IsOk() )
+    {
+        prevItem = GetItemParent(item);
+    }
+
+    // find the first visible item after it
+    while ( prevItem.IsOk() && !IsVisible(prevItem) )
+    {
+        prevItem = GetNext(prevItem);
+        if ( !prevItem.IsOk() || prevItem == item )
+        {
+            // there are no visible items before item
+            return wxTreeItemId();
+        }
+    }
+
+    // from there we must be able to navigate until this item
+    while ( prevItem.IsOk() )
+    {
+        const wxTreeItemId nextItem = GetNextVisible(prevItem);
+        if ( !nextItem.IsOk() || nextItem == item )
+            break;
+
+        prevItem = nextItem;
+    }
+
+    return prevItem;
 }
 
 wxTreeItemId wxTreeCtrl::AddRoot(const wxString& text,
                              int image, int selImage,
                              wxTreeItemData *data)
 {
-    return wxTreeItemId();
+    QTreeWidgetItem *root = m_qtTreeWidget->invisibleRootItem();
+    wxTreeItemId newItem = DoInsertItem(
+        wxQtConvertTreeItem(root),
+        0,
+        text,
+        image,
+        selImage,
+        data
+    );
+
+    m_qtTreeWidget->setCurrentItem(NULL);
+
+    if ( (GetWindowStyleFlag() & wxTR_HIDE_ROOT) != 0 )
+        m_qtTreeWidget->setRootIndex(m_qtTreeWidget->model()->index(0, 0));
+    else
+        m_qtTreeWidget->setRootIndex(QModelIndex());
+
+    return newItem;
 }
 
 void wxTreeCtrl::Delete(const wxTreeItemId& item)
 {
+    wxCHECK_RET(item.IsOk(), "invalid tree item");
+
+    QTreeWidgetItem *qTreeItem = wxQtConvertTreeItem(item);
+    QTreeWidgetItem *parent = qTreeItem->parent();
+
+    if ( parent != NULL )
+    {
+        parent->removeChild(qTreeItem);
+    }
+    else
+    {
+        m_qtTreeWidget->removeItemWidget(qTreeItem, 0);
+    }
+
```

Comment:
It seems strange that `Delete()` doesn't call `DeleteChildren()`. Maybe this happens implicitly somehow, but normally `Delete() == DeleteChildren() + DeleteSelf()`. Does it really work like this here?

Reply:
I'm going to look into this now.   The QT docs are, as always, unclear about whether it handles the delete recursively. 

If it does, do we just need to generate the appropriate events for the children being deleted?

Reply:
Have modified delete to call `DeleteChildren()` for the item being deleted.

Reply:
I _think_ we do get events under MSW currently, but ideal would be to add a unit test checking for this.

Reply:
As part of this commit, I modified  the Delete test to include a child item and updated the assertion to check for 2 items being deleted.  

Is this enough or did you have a new test case in mind?